### PR TITLE
merkle tree impl

### PR DESCRIPTION
**Pull request by jobez**
Merkle implementation needed for FRI

### CODE REVIEW

File: src/iop/types/merkle.rs [hunk 1]
```diff
@@ -0,0 +1,283 @@
...
+}
+
+impl From<[u8; 64]> for MerkleHash {
+    fn from(inner: [u8; 64]) -> Self {
+        Self { inner }
+    }
+}
+
+impl MerkleHash {
+    pub fn as_bytes(&self) -> &[u8; 64] {
```
Comment (greged93) [inline src/iop/types/merkle.rs:22]: Does this work? I thought you couldn't return a reference from the function itself?

File: src/iop/types/merkle.rs [hunk 2]
```diff
@@ -0,0 +1,283 @@
...
+        assert!(leaves.len().is_power_of_two());
+
+        let mut hashes: Vec<MerkleHash> = leaves.to_vec();
+
+        // A buffer is used to store the results of each round of hashing.
+        // By allocating the buffer once and reusing it, we avoid
+        // the overhead of allocating a new vector in each iteration.
+        // Using clear() just resets the length of the vector without
+        // deallocating the memory, making it more efficient for our purposes.
+        let mut buffer = Vec::with_capacity(hashes.len() / 2);
```
Comment (greged93) [inline src/iop/types/merkle.rs:45]: that's dope

File: src/iop/types/merkle.rs [hunk 3]
```diff
@@ -0,0 +1,283 @@
...
+
+            for pair in hashes.chunks(2) {
+                hasher.update(pair[0].as_bytes());
+                hasher.update(pair[1].as_bytes());
+                let result: [u8; 64] = hasher.finalize_reset().into();
+                buffer.push(MerkleHash::from(result));
+            }
+            // Swapping the references of hashes and buffer avoids
+            // copying the data and prepares for the next round.
+            std::mem::swap(&mut buffer, &mut hashes);
```
Comment (greged93) [inline src/iop/types/merkle.rs:59]: I didn't know you could do this, so this transfers the data from one vector to the other?

File: src/iop/types/merkle.rs [hunk 4]
```diff
@@ -0,0 +1,283 @@
...
+        }
+
+        // Add the sibling of our desired leaf to the results.
+        results.push(current_slice[1 - current_index]);
+        // Reverse the order of results to match the path from leaf to root.
+        results.reverse();
+        results
+    }
+
+    pub fn verify(&self, leaf: MerkleHash, path: &[MerkleHash], index: usize) -> bool {
```
Comment (greged93) [inline src/iop/types/merkle.rs:110]: is there a reason to have verify and is_leaf_in_tree public? seems like one should be private

File: src/iop/types/merkle.rs (hunk 3 above)
Comment (jobez) [reply src/iop/types/merkle.rs:59]: yep: `hashes` will contain the results of the most recent round of hashing, and `buffer` will be empty, ready to be reused for the next round

File: src/iop/types/merkle.rs (hunk 1 above)
Comment (jobez) [reply src/iop/types/merkle.rs:22]: you can return references to the internal data of a struct as long as there's lifetime overlap

the ownership by MerkleHash always 'covers' the inner reference, if that makes sense.

File: src/iop/types/merkle.rs (hunk 4 above)
Comment (jobez) [reply src/iop/types/merkle.rs:110]: can bring is_leaf_in_tree back into privacy!

**Pull request merged**